File: recommand-python/com/music/recommand/util/predict.py
import numpy


class Predict:

    x = None
    sim = None
    mean_rate = None
    sorted_sim_arg = None

    def __init__(self, x, sim):
        self.x = x
        self.sim = sim
        self.sorted_sim_arg = sim.argsort()[:, ::-1]
        self.mean_rate = x.mean(axis=1)

    def predict(self, u, k=50):
        sorted_sim_arg = self.sorted_sim_arg[u, 0:k]
        sorted_sim = self.sim[u, sorted_sim_arg][:, 0:k]
        mu = self.mean_rate[u]
        if sorted_sim.sum() != 0:
            mean = self.mean_rate[sorted_sim_arg][0]
            index = sorted_sim_arg.tolist()[0]
            rate = self.x[index]
            delta = rate - mean
            p = mu + sorted_sim * delta / sorted_sim.sum()
        else:
            p = numpy.zeros((1, self.x.shape[1]))
        return p

File: matlab/predict.m
function p = predict(x, sim, u, k)

[st, index] = sort(sim, 2, 'descend');
simk = st(u, 1:k); % 前 k 个临近用户的相似度
idxk = index(u, 1:k);   % 前 k 个临近用户的id
m = mean(x, 2);     % 所有用户评分的平均值
mu = m(u, :);       % 目标用户评分的平均值

if sum(simk) ~= 0
    p = mu + simk * (x(idxk, :) - m(idxk)) / sum(simk);
else
    p = zeros(1, size(x, 2));
end
fprintf("predict %d finished!\n", u);
end